### ISSUE TITLE
FR: allow loading phase failures to be caught in Starlark tests

### ISSUE THREAD

**Issue by tetromino** [bazelbuild/bazel#18008]
Currently Skylib has [loadingtest.equals](https://github.com/bazelbuild/bazel-skylib/blob/main/lib/unittest.bzl#L636) to verify that a macro has expected output. However, we have no way to verify that a macro *fails* as expected: a loading phase failure cannot be caught, it causes the build to fail. Thus there is no loading phase analogue of `analysistest.make(..., expect_failure = True)`.

It would be useful if we had a mechanism similar to `--allow_analysis_failures` to make it possible to write a Starlark test framework that can catch and verify loading phase tests.

Potentially interested parties: @murali42

**Comment by brandjon** [comment]
Discussed offline. The prototype implementation for this FR adds a new builtin on the `native` object that calls a given arbitrary callable and intercepts `EvalException` for inspection by Starlark code. This is a quite powerful API in that it essentially grants exception catching capability to the loading phase (where `fail()` is equivalent to raise/throw).

We wouldn't want to casually add such a feature without considerable design discussion. Exception handling has lots of risks: more interpreter logic, more cases for tooling to handle, and most importantly, more implicit control flow that hurts readability. Today, there's no reason for Starlark code to care what happens once a `fail()` statement is reached. But with this feature, any caller of `fail()`, including internally raised `EvalExceptions`, have to consider how the build may correctly proceed without error (or crash) in spite of the failure.

At the Starlark language level, we'd want a community-wide discussion before adding this capability. At the Bazel level, we'd still want a design doc and lots of discussion. There's also the question of whether loading-time testing should be enabled with a FR like this one or via some alternative mechanism, e.g. productionizing the method for running Bazel within Bazel like we do in our integration tests.

One alternative I'd propose off the top of my head: Don't add a new Starlark builtin, but create a way to evaluate a file or a source code text string as if it were a package, and assert on the resulting error strings. Then invoke this from a new native-defined "macro_test" rule. Again, we'd need design discussion, and I don't think improved testing support is on our immediate roadmap.

**Comment by Wyverald** [comment]
> Exception handling has lots of risks: more interpreter logic, more cases for tooling to handle, and most importantly, more implicit control flow that hurts readability.

Assuming that we're still talking about a new builtin `native.assert_failure(lambda)` whose implementation calls `lambda` and catches any thrown `EvalException`, I'm confused what these risks are:

- What extra interpreter logic is there? We're simply catching the exception in the body of this specific builtin, *not* adding exception handling capabilities to arbitrary Starlark code.
- What extra cases do you expect tooling to handle?
- This cannot be used to implement implicit control flow unless you abuse `native.assert_failure()`.

> Today, there's no reason for Starlark code to care what happens once a fail() statement is reached. But with this feature, any caller of fail(), including internally raised EvalExceptions, have to consider how the build may correctly proceed without error (or crash) in spite of the failure.

This for me is a much more convincing point. Intuitively it feels like loading-phase code cannot affect the resultant package in any other way than declaring a new target, and declaring a new target (or failing to do so) shouldn't affect other parts of the package (other than stuff like `native.existing_rules()`). But we may need to audit the package building code to ensure there are no half-baked `Rule` instances being inserted into a `Package`.

---

In summary I think the trepidation may be overblown, especially considering that we're not adding `try-catch` to Starlark, but a very limited built-in that just catches loading-phase eval errors.

**Comment by rickeylev** [comment]
I just wanted to drop an alternative idea inspired by testing.analysis_test. Basically, a custom rule takes a setup function as an argument, e.g. something like:

```
def _test_bad_value_rejected_setup(...):
  my_rule(
    name = "bad_target",
    foo = "python2"
  )

def _test_bad_value_rejected_impl(<something probably ctx-like>):
  <code to assert "bad foo value: python2" in ctx.failures>
 
testing.loading_failure_test(
  name = "test_bad_value_rejected",
  setup = _test_bad_value_rejected_setup,
  impl = _test_bad_value_rejected_impl,
)
```

Internally, it calls the setup function, captures any errors, and then passes on that information to the assertion function (the assertion function could itself be basically modeled as a rule invocation, like testing.analysis_test does). If it takes additional arguments, like the build language flag settings, or something to control the environment (e.g., something to make it look like a BUILD vs bzl vs WORKSPACE vs MODULE runtime environment), then...well, maybe that helps provide a way to test those cases.

----

> a way to evaluate a file or a source code text

This would be fantastic and provide a nice solution to several problems.

 * Testing bzlmod modules. This is very hard/impossible right now. It'd be much easier if you could
   provide a fake file system as input. The same is probably true for testing other
   things like WORKSPACE behavior or repository rules.
 * Testing "examples". Something I've noticed in Bazel rule sets is they have a lot of
   example sub-projects. These not only serve as examples, though, but also as a way to
   test specific WORKSPACE/MODULE usages. They're usually pretty simple, but end up
   requiring a variety of surrounding infrastructure to run correctly (git ignore files,
   bazel ignore files, --deleted_packages flag settings, more bazelrcs, additional
   CI confguration). These examples also tend to grow extra pieces to test additional
   cases because adding an example involves some toil.
 * Provide a straight-forward migration path for all the Java-based tests in Bazel
   for rules that are now implemented in Starlark.

**Comment by tetromino** [comment]
Issue is still relevant

**Comment by jjmaestro** [comment]
Landing here after posting on Slack ([see 🧵](https://bazelbuild.slack.com/archives/CA31HN1T3/p1740583086685719)). @fmeum pointed at the alternative that @rickeylev proposed. While I think that's OK, I would really love to have a simple `native.assert_failure()`.

And like @Wyverald I agree that the restricted case of just having `native.assert_failure()` would mean not adding `try-catch` to Starlark, and thus it would hopefully be much easier and/or require less discussions.

Hopefully someone with expertise on Bazel internals can quickly code this! 😅  Also, @brandjon you said

> Discussed offline. The prototype implementation for this FR adds a new builtin on the native object that calls a given arbitrary callable and intercepts EvalException for inspection by Starlark code. This is a quite powerful API in that it essentially grants exception catching capability to the loading phase (where fail() is equivalent to raise/throw).

So now, I'm wondering if that prototype PR would be the code that's needed for this? That is, not a complete implementation of `try-catch` in Bazel Starlark but a concrete function in the `native` object :-?  I'd love to see if that's the case and we could get a PR out of it!

Thanks!!

**Comment by fmeum** [comment]
Still relevant
